examples/vision/retinanet.py

```diff
@@ -0,0 +1,700 @@
+        box_predictions = predictions[:, :, :4]
+        cls_predictions = tf.nn.sigmoid(predictions[:, :, 4:])
+        boxes = self._decode_box_predictions(anchor_boxes[None, ...], box_predictions)
+        boxes = tf.expand_dims(boxes, axis=2)
+        boxes = tf.tile(boxes, [1, 1, self.num_classes, 1])
+
+        return tf.image.combined_non_max_suppression(
+            boxes,
+            cls_predictions,
+            self.max_detections_per_class,
+            self.max_detections,
+            self.nms_iou_threshold,
+            self.confidence_threshold,
+            clip_boxes=False,
+        )
+
+
+"""
+## Implementing Smooth L1 loss and Focal Loss as keras custom losses
+"""
+
+
+class RetinaNetBoxLoss(tf.losses.Loss):
+    """Implements Smooth L1 loss"""
+
+    def __init__(self, delta):
+        super(RetinaNetBoxLoss, self).__init__(
+            reduction="none", name="RetinaNetBoxLoss"
+        )
+        self._delta = delta
+
+    def call(self, y_true, y_pred):
+        loss = tf.compat.v1.losses.huber_loss(
+            labels=y_true, predictions=y_pred, delta=self._delta, reduction="none"
+        )
+        loss = tf.reduce_sum(loss, axis=-1)
+        return loss
+
+
+class RetinaNetClassificationLoss(tf.losses.Loss):
+    """Implements Focal loss"""
+
+    def __init__(self, alpha, gamma):
+        super(RetinaNetClassificationLoss, self).__init__(
+            reduction="none", name="RetinaNetClassificationLoss"
+        )
+        self._alpha = alpha
+        self._gamma = gamma
+
+    def call(self, y_true, y_pred):
+        cross_entropy = tf.nn.sigmoid_cross_entropy_with_logits(
+            labels=y_true, logits=y_pred
+        )
+        probs = tf.nn.sigmoid(y_pred)
+        alpha = tf.where(tf.equal(y_true, 1.0), self._alpha, (1.0 - self._alpha))
+        pt = tf.where(tf.equal(y_true, 1.0), probs, 1 - probs)
+        loss = alpha * tf.pow(1.0 - pt, self._gamma) * cross_entropy
+        return tf.reduce_sum(loss, axis=-1)
+
+
+class RetinaNetLoss(tf.losses.Loss):
+    """Wrapper to combine both the losses"""
+
+    def __init__(self, num_classes=80, alpha=0.25, gamma=2.0, delta=1.0):
+        super(RetinaNetLoss, self).__init__(reduction="auto", name="RetinaNetLoss")
+        self._cls_loss = RetinaNetClassificationLoss(alpha, gamma)
+        self._box_loss = RetinaNetBoxLoss(delta)
+        self._num_classes = num_classes
+
+    def call(self, y_true, y_pred):
+        y_pred = tf.cast(y_pred, dtype=tf.float32)
+        box_labels = y_true[:, :, :4]
+        box_predictions = y_pred[:, :, :4]
+        cls_labels = tf.one_hot(
+            tf.cast(y_true[:, :, 4], dtype=tf.int32),
+            depth=self._num_classes,
+            dtype=tf.float32,
+        )
+        cls_predictions = y_pred[:, :, 4:]
+        positive_mask = tf.cast(tf.greater(y_true[:, :, 4], -1.0), dtype=tf.float32)
+        ignore_mask = tf.cast(tf.equal(y_true[:, :, 4], -2.0), dtype=tf.float32)
```

Comment:
why is ignore_mask participating in final loss? Shouldn't it be positive and negative mask?

Reply:
Oh nvm, it looks like you're excluding them from cls loss